lib/dao.js

```diff
@@ -1526,6 +1526,27 @@ function NumberType(val) {
   return !isNaN(num) ? num : val;
 }
 
+function coerceArray(val) {
+  if (Array.isArray(val)) {
+    return val;
+  }
+
+  if (!utils.isPlainObject(val)) {
+    throw new Error(g.f('Value is not an {{array}} or {{array-like object}}'));
+  }
+
+  var arrayVal = new Array(Object.keys(val).length);
```

Comment:
Hey @bajtos :) So, there are two aspects here, the code's behavior and parity with qs, both of which I'll address.

Regarding the code from this PR - I believe that this code in fact specifically avoids the issue you're concerned about, but in the very least, the fact that this is not clear means that something needs to change here. That said, let me decompose the code a bit to explain it.

The code operates on an assumption that a well-formed "array-like object" contains a series of sequential numeric indices starting from zero. It initializes the `arrayVal` array with the length of the keys from the source content (`Object.keys(arg).length`, which is not an arbitrarily large number), and is very careful to never perform an assignment operation like that from your example (`result[4000000000]=true`).

To avoid assignments like that, it iterates from 0 up until the keys length, and hard fails in case of any missing or unexpected keys. In addition to protecting against the arbitrary-sized key, it also rejects any objects that are cluttered with unexpected non-sequential non-numeric keys.  Does this make sense?

Now, regarding qs's approach - yeah generally I agree that it's better to avoid the potential for subtle discrepancies. However, they seem to have decided to entirely disable this array parsing by default in a future release, meaning even single element array-like parameters would be represented as objects. The premise behind this PR was to anticipate that upcoming change, and my implementation aims to provide a more conservative approach to array processing than `qs` is currently performing. 

So it seems like there are several routes to go here regarding qs's arrayLimit,
   1. Do nothing, however this will be a decision to (in the future) disable by default support for array-like parameters (`filter[where][someprop][inq][0]=...`).
   2. Pick more a reasonable default for `arrayLimit` as part of the default Loopback config. Currently it defaults to 20, which is easy to unexpectedly stumble into (potentially disastrously). A substantially larger value (1000?) would create no practical risk and likely avoid the problem in almost all cases.
   3. A patch like the one from this PR, which seeks to provide support for array-like objects in the future.

Thoughts?